hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java

```diff
@@ -451,25 +451,26 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
             progress.cancel();
             return false;
           }
-          if (kvs != null && bytesWrittenProgressForShippedCall > shippedCallSizeLimit) {
-            if (lastCleanCell != null) {
-              // HBASE-16931, set back sequence id to avoid affecting scan order unexpectedly.
-              // ShipperListener will do a clone of the last cells it refer, so need to set back
-              // sequence id before ShipperListener.beforeShipped
-              PrivateCellUtil.setSequenceId(lastCleanCell, lastCleanCellSeqId);
-            }
-            // Clone the cells that are in the writer so that they are freed of references,
-            // if they are holding any.
-            ((ShipperListener) writer).beforeShipped();
-            // The SHARED block references, being read for compaction, will be kept in prevBlocks
-            // list(See HFileScannerImpl#prevBlocks). In case of scan flow, after each set of cells
-            // being returned to client, we will call shipped() which can clear this list. Here by
-            // we are doing the similar thing. In between the compaction (after every N cells
-            // written with collective size of 'shippedCallSizeLimit') we will call shipped which
-            // may clear prevBlocks list.
-            kvs.shipped();
-            bytesWrittenProgressForShippedCall = 0;
+        }
+        if (kvs != null && bytesWrittenProgressForShippedCall > shippedCallSizeLimit) {
+          if (lastCleanCell != null) {
+            // HBASE-16931, set back sequence id to avoid affecting scan order unexpectedly.
+            // ShipperListener will do a clone of the last cells it refer, so need to set back
+            // sequence id before ShipperListener.beforeShipped
+            PrivateCellUtil.setSequenceId(lastCleanCell, lastCleanCellSeqId);
+            lastCleanCell = null;
```

Comment:
What I am saying is this set to null is enough to solve the problem you saw.
You do not have to move the block outside of for loop. So  this single line change will fix it.  You have a repro. Will u be able to confirm pls?   We can avoid larger change/

Reply:
Do you mean
`        
for (Cell c : cells) {
          writer.append(c);
          if (kvs != null && bytesWrittenProgressForShippedCall > shippedCallSizeLimit) {
            if (lastCleanCell != null) {
              PrivateCellUtil.setSequenceId(lastCleanCell, lastCleanCellSeqId);
              lastCleanCell = null;
            }
            ((ShipperListener) writer).beforeShipped();
            kvs.shipped();
          }
}
`?
 I do not think this can fix this issus.

Reply:
yes.
Before calling shipped (which will release BBs) we are setting back the seqId.   The beforeShipped call make sure that other layers which hold cell ref will copy those cells.  Now after reset the seqId, we make lastCleanCell  = null.  So no chance of calling again setSequenceId on same Cell.  If u have a repro, will u be able to check pls?

Reply:
With this fix(Only set lastCleanCell = null), the UT is failed too.

The main problem is not related to lastCell. 
It is when `writer.append(c)`, the c is aleady released by previous `kvs.shipped()`.

Reply:
I see it now.. You have a point.  +1 then.

Reply:
BTW excellent investigation work done to pin point what is the issue.